### ISSUE TITLE
New key, right_alt would be appreciate :)

### ISSUE THREAD

**Issue by hadrien-toma** [octalmage/robotjs#234]
Hello there, great job in this repo!
Just an issue to mention that I would appreciate the use of the missing key "right_alt".

**Comment by hadrien-toma** [comment]
In fact, I miss several other keycodes, could you explain me how to add one? I will contribute for the others and submit my pull requests.

**Comment by octalmage** [comment]
Hi! Glad you like it. 

You can add keycodes here:

https://github.com/octalmage/robotjs/blob/master/src/keycode.h#L58

There's an enum for each OS. Then it will need to be added here:

https://github.com/octalmage/robotjs/blob/master/src/robotjs.cc#L336

Let me know if you need any help!

**Comment by astrid-at-heart** [comment]
Looks like it's just `kVK_RightOption` `vk_rightalt` and `k_rightalt` according to many platforms  
I only doubt the RightOption, images tell me alt and option share a key, but some images uses an icon instead

If you're looking for more keycodes
https://github.com/weeeBox/oldies-xna-framework/blob/4c1a19552539a6a497c1071e9d8957dd34651804/trunk/Framework/AsapXna/asap/core/KeyCode.cs#L152

https://github.com/soegaard/remacs/blob/1b184add294a8ff5aa55ea30b82eb1e3bf91e643/osx-keyboard.rkt

https://gist.github.com/JohnJB/5662272

**Comment by Scrxtchy** [comment]
Looks like it's just `kVK_RightOption` `vk_rightalt` and `k_rightalt` according to many platforms  
I only doubt the RightOption, images tell me alt and option share a key, but some images uses an icon instead

If you're looking for more keycodes
https://github.com/weeeBox/oldies-xna-framework/blob/4c1a19552539a6a497c1071e9d8957dd34651804/trunk/Framework/AsapXna/asap/core/KeyCode.cs#L152

https://github.com/soegaard/remacs/blob/1b184add294a8ff5aa55ea30b82eb1e3bf91e643/osx-keyboard.rkt

https://gist.github.com/JohnJB/5662272

**Comment by hadrien-toma** [comment]
Thank you both for your help!

This is a [spreadhseet](https://docs.google.com/spreadsheets/d/1gzaQugJOGgD9pxyT4EkFX3r4pmcBe6bio2jwCbBFu_g) containing the information you gave me.
I am wondering if there are no conflit between them. I left empty the Cells that are not valid assuming the policy I guess for this project. What brings me to think that these empty cells may be conflicting if I fill them? For two unlinked reasons :
* About the Windows column of the RobotJS sheet : The keys I found from the link given by @Scrxtchy are not the same than the ones I found in this file https://github.com/octalmage/robotjs/blob/master/src/keycode.h#L58 :
    * For instance for K_PAGEUP :

**Comment by Scrxtchy** [comment]
Ahh, so it seems
This might be a bit more accurate

https://github.com/me2d13/luamacros/blob/eee57bddff37396345173c647240982f526932e5/src/uSendKeys.pas
This comes out of software I actually use

**Comment by hadrien-toma** [comment]
Okay thank you, I'll check. The case is always conflicting me, should I use upper or not?

**Comment by Scrxtchy** [comment]
The `fSequence[I]` case?
don't worry about it, it's for the software to handle pressing shift or modifiers alongside a key, like an autohotkey script
```
  elseif (button == 106) then lmc_send_keys('%{F24}') --Astericks
  elseif (button == 111) then lmc_send_keys('^{F23}') --forward slash   
```

**Comment by hadrien-toma** [comment]
Hum... What I am wondering is : Can I use [these keycodes](https://github.com/me2d13/luamacros/blob/eee57bddff37396345173c647240982f526932e5/src/uSendKeys.pas#L129-L141) in RobotJS keycode.h in upper case?

Any idea of a link where I can find XF86XK_* keys ? :D I am looking forward to begin to code^^.

**Comment by Scrxtchy** [comment]
I am unsure really, these are added manually due to them not being in the library, so instead it's providing the number used, instead of the reference variable, thus the inconstancy. As as pascal is a non-case sensitive language, as these are that freedom exists for future referral in the application

**Comment by hadrien-toma** [comment]
Okay, I'll try and tell you if builds work^^.

**Comment by Scrxtchy** [comment]
If I could bother to set up gyp for actually building anything now I could hammer it out myself as well and see what works
But it's too much of a hassle

**Comment by hadrien-toma** [comment]
Haha you are right!

**Comment by Scrxtchy** [comment]
Have you seen this list of XF86XK_* keys?
https://cgit.freedesktop.org/xorg/proto/x11proto/tree/keysymdef.h
https://cgit.freedesktop.org/xorg/proto/x11proto/tree/XF86keysym.h

**Comment by hadrien-toma** [comment]
Hey,

So I reach to build the code from files :
* [keycode.h](https://github.com/octalmage/robotjs/files/466101/keycode.h.txt)
* [robotjs.cc](https://github.com/octalmage/robotjs/files/466102/robotjs.cc.txt)

**But when I run this :**
`var robot = require('../local_packages/robotjs/build/Release/robotjs.node')
robot.keyTap("tild")
robot.keyTap("control_right")
`

**This is the error I have :**
robot.keyTap("control_right")
      ^
Error: Invalid key code specified.
    at Error (native)
    at Object.<anonymous> (/home/hadrien-toma/dev/nodejs/test/index.js:5:7)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:458:32)
    at tryModuleLoad (module.js:417:12)
    at Function.Module._load (module.js:409:3)
    at Module.runMain (module.js:575:10)
    at run (bootstrap_node.js:352:7)
    at startup (bootstrap_node.js:144:9)

**It seems to be the same for all keys I added. An idea of what I have to do next?**

**Comment by octalmage** [comment]
Hi @hadrien-toma,

Did you rebuild the binary with node-gyp? The source fils look correct.

**Comment by hadrien-toma** [comment]
My build line is :
`rm -rf ./build ; rm -rf node_modules ; npm install`
I think this triggers node-gyp, doesn't it?

**Comment by octalmage** [comment]
By default npm install will download the pre-built binaries. You'll need to run 

    node-gyp rebuild

In node_modules/robotjs/

**Comment by Scrxtchy** [comment]
Would any modifications of this nature be open for pull requests?

**Comment by hadrien-toma** [comment]
I will emit a pull request as soon as I have tested my changes to not commit untested code but if someone else wants to test it for us (and tell us the result^^), my changes are in the 2 abovementioned files, I will thank him for that.

**Comment by jonalbertini** [comment]
still nothing ? I've spent some time on it but it seems that no keyboard code fit for the right alt (AltGr).
after editing the files as you submitted I  just can't compile anything ... some binding.gyp error ... I just can't spend some more time on it by now . but maybe will soon. 
great job by the way, it's one of the best automation module I've see ! 
(ps: I've met some trouble with the € thing also... but maybe related, it uses the AltGr key to type)

**Comment by hadrien-toma** [comment]
Unfortunately, I don't see how to fix neither #235 nor #16, so I still can't use my modifications :( ... If you reach to something later, I will thank you to let us know! ;)

**Comment by oktapodia** [comment]
This issue has been closed because it has been inactive for > 180 days.
You may reopen this issue if it has been closed in error.
